collector/transport_zone_collector.go

```diff
@@ -0,0 +1,136 @@
+		prometheus.BuildFQName(namespace, "transport_zone", "down_transport_node"),
+		"Total number of transport node with down status in transport zone",
+		[]string{"id", "name"},
+		nil,
+	)
+	transportZoneUnknownNode := prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "transport_zone", "unknown_transport_node"),
+		"Total number of transport node with unknown status in transport zone",
+		[]string{"id", "name"},
+		nil,
+	)
+	transportZoneUpNode := prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "transport_zone", "up_transport_node"),
+		"Total number of transport node with up status in transport zone",
+		[]string{"id", "name"},
+		nil,
+	)
+	return &transportZoneCollector{
+		transportZoneClient:        nsxtClient,
+		logger:                     logger,
+		transportZoneLogicalPort:   transportZoneLogicalPort,
+		transportZoneLogicalSwitch: transportZoneLogicalSwitch,
+		transportZoneTransportNode: transportZoneTransportNode,
+		transportZoneDegradedNode:  transportZoneDegradedNode,
+		transportZoneDownNode:      transportZoneDownNode,
+		transportZoneUnknownNode:   transportZoneUnknownNode,
+		transportZoneUpNode:        transportZoneUpNode,
+	}
+}
+
+// Describe implements the prometheus.Collector interface.
+func (c *transportZoneCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- c.transportZoneLogicalPort
+	ch <- c.transportZoneLogicalSwitch
+	ch <- c.transportZoneTransportNode
+	ch <- c.transportZoneDegradedNode
+	ch <- c.transportZoneDownNode
+	ch <- c.transportZoneUnknownNode
+	ch <- c.transportZoneUpNode
+}
+
+// Collect implements the prometheus.Collector interface.
+func (c *transportZoneCollector) Collect(ch chan<- prometheus.Metric) {
+	transportZones, err := c.transportZoneClient.ListAllTransportZones()
+	if err != nil {
+		level.Error(c.logger).Log("msg", "Unable to list transport zones", "err", err)
+		return
+	}
+	c.collectTransportZonesStatus(transportZones, ch)
+	c.collectTransportZonesHeatmapStatus(transportZones, ch)
+}
+
+func (c *transportZoneCollector) collectTransportZonesStatus(transportZones []manager.TransportZone, ch chan<- prometheus.Metric) {
+	for _, transportZone := range transportZones {
+		transportZoneStatus, err := c.transportZoneClient.GetTransportZoneStatus(transportZone.Id)
+		if err != nil {
+			level.Error(c.logger).Log("msg", "Unable to get transport zone status", "id", transportZone.Id, "err", err)
+			continue
+		}
+		transportZoneLabels := []string{transportZone.Id, transportZone.DisplayName}
+		ch <- prometheus.MustNewConstMetric(c.transportZoneLogicalPort, prometheus.GaugeValue, float64(transportZoneStatus.NumLogicalPorts), transportZoneLabels...)
+		ch <- prometheus.MustNewConstMetric(c.transportZoneLogicalSwitch, prometheus.GaugeValue, float64(transportZoneStatus.NumLogicalSwitches), transportZoneLabels...)
+		ch <- prometheus.MustNewConstMetric(c.transportZoneTransportNode, prometheus.GaugeValue, float64(transportZoneStatus.NumTransportNodes), transportZoneLabels...)
```

Comment:
how does transportZoneTransportNode different from transportNode in transportNodeCollector?

Reply:
Transport node is the same, the problem here is that the connection between transport zone and transport node is many to many.
Is there's a recommendation on connecting both metrics?